### PR TITLE
✨ Improve Google matched content logic to reflect adsbygoogle.js.

### PR DESCRIPTION
**Pull request by nbeloglazov**
The commit introduces core-shared.js file which supposed to be shared between AMP and internal Google adsbygoogle.js codebase. The file has minimal dependencies on environment to make keeping them in sync easier.

In particular this commit adds support for publisher settings as described in https://support.google.com/adsense/answer/7533385

Tested:
  unit tests
  manually on fast fetch and delayed fetch

### CODE REVIEW

File: ads/google/core-shared.js [hunk 1]
```diff
@@ -0,0 +1,482 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
```
Comment (torch2424) [inline ads/google/core-shared.js]: Incorrect copyright year 😄

File: ads/google/core-shared.js [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+  };
+}
+
+/**
+ * Calculate the ad height according to the layout and ad width.
+ * @param {number} adWidth
+ * @param {!LayoutType} layout
+ * @return {number}
+ */
+const getAdHeight = (adWidth, layout) => {
```
Comment (lannka) [inline ads/google/core-shared.js]: consistency: shall we use `function`?

File: ads/google/core-shared.js [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+  'data-matched-content-columns-num=\"1,6\"\n' +
+  'data-matched-content-ui-type=\"image_stacked,image_card_sidebyside\"';
+
+/**
+ * Configuration of content recommendation unit for current slot. This is the
+ * result of running CoRe responsive logic and values from this config
+ * will be used in ad request.
+ * @record
+ */
+class CoReConfig { // eslint-disable-line no-unused-vars
```
Comment (lannka) [inline ads/google/core-shared.js]: Does a TypeDef work here? Is a class needed?

File: ads/google/core-shared.js [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+    paramsInvalid: false,
+  };
+}
+
+/**
+ * @param {number} availableWidth
+ * @param {boolean} isMobile
+ * @return {!CoReConfig}
+ */
+export function getAutoConfig(availableWidth, isMobile) {
```
Comment (lannka) [inline ads/google/core-shared.js]: please organize the file so that public function is before local function.

File: ads/google/core-shared.js [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+const MAX_PUB_CONTROL_DIMENSION = 1500;
+
+/**
+ * Minimum width of desktop responsive slot in CoRe responsive. We have
+ * different logic for desktop and mobile slots. Any slot width equal or larger
+ * than this will be adapted to the desktop logic while any slot width smaller
+ * than this will be adapted to the mobile logic.
+ * @const {number}
+ */
+export const MIN_PUB_CONTROL_WIDTH_OF_DESKTOP = 468;
```
Comment (lannka) [inline ads/google/core-shared.js]: please move public constants before local ones

File: ads/google/core-shared.js (hunk 1 above)
Comment (nbeloglazov) [reply ads/google/core-shared.js]: Thanks, done.

File: ads/google/core-shared.js (hunk 4 above)
Comment (nbeloglazov) [reply ads/google/core-shared.js]: Done.

File: ads/google/core-shared.js (hunk 3 above)
Comment (nbeloglazov) [reply ads/google/core-shared.js]: Generally styleguide recommends records over typedefs: https://google.github.io/styleguide/jsguide.html#jsdoc-enum-and-typedef-comments

File: ads/google/core-shared.js (hunk 2 above)
Comment (nbeloglazov) [reply ads/google/core-shared.js]: Done.

File: ads/google/core-shared.js (hunk 5 above)
Comment (nbeloglazov) [reply ads/google/core-shared.js]: Done.

**Comment by nbeloglazov** [comment]
Addressed comments. And refactored code to not depend on any other AMP libraries as the code will be imported and used by internal google tags. Please take another look. Thanks!

File: ads/google/core-shared.js [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+      layoutType: LayoutType.IMAGE_STACKED,
+    };
+  }
+}
+
+/**
+ * Get CoRe Pub Control UI Sizes.
+ * @param {number} availableWidth
+ * @param {{
+ *   numberOfRows: (string|undefined),
```
Comment (keithwrightbos) [inline ads/google/core-shared.js]: You repeat this type multiple times, can you make it a proper typedef?

Comment (nbeloglazov) [reply ads/google/core-shared.js]: Done. Introduced RawPublisherControlParams typedef.

**Pull request merged**